h/form.py

```diff
@@ -89,6 +95,78 @@ def configure_environment(config):  # pragma: no cover
     config.registry[ENVIRONMENT_KEY] = create_environment(base)
 
 
+def handle_form_submission(request, form, on_success, on_failure):
+    """
+    Handle the submission of the given form in a standard way.
+
+    :param request: the Pyramid request
+
+    :param form: the form that was submitted
+    :type form: deform.form.Form
+
+    :param on_success:
+        A callback function to be called if the form validates successfully.
+
+        This function should carry out the action that the form submission
+        requests (for example for a change password form, this function would
+        change the user's password) and return the view callable result that
+        should be returned if this is not an XHR request.
+
+        If on_success() returns ``None`` then ``handle_form_submission()``
+        will return ``HTTPFound(location=request.url)`` by default.
+    :type on_success: callable
+
+    :param on_failure:
+        A callback function that will be called if form validation fails in
+        order to get the view callable result that should be returned if this is
+        not an XHR request.
+    :type on_failure: callable
+
+    """
+    try:
+        appstruct = form.validate(request.POST.items())
+    except deform.ValidationFailure:
+        result = on_failure()
+    else:
+        result = on_success(appstruct)
+
+        if result is None:
+            result = httpexceptions.HTTPFound(location=request.url)
+
+        request.session.flash(_("Success. We've saved your changes."),
+                              'success')
+
+    return to_xhr_response(request, result, form)
+
+
+def to_xhr_response(request, non_xhr_result, form):
+    """
+    Return an XHR response for the given ``form``, or ``non_xhr_result``.
+
+    If the given ``request`` is an XMLHttpRequest then return an XHR form
+    submission response for the given form (contains only the ``<form>``
+    element as an HTML snippet, not the entire HTML page).
+
+    If ``request`` is not an XHR request then return ``non_xhr_result``, which
+    should be the result that the view callable would normally return if this
+    were not an XHR request.
+
+    :param request: the Pyramid request
+
+    :param non_xhr_result: the view callable result that should be returned if
+        ``request`` is *not* an XHR request
+
+    :param form: the form that was submitted
+    :type form: deform.form.Form
+
+    """
+    if not request.is_xhr:
+        return non_xhr_result
+
+    request.override_renderer = 'string'
+    return form.render()
```

Comment:
Since no argument is being passed to `render()`, this will return the form without any of the fields filled in. If we naively apply the result on the client (eg. with `formElement.innerHTML = <updated form>`) this will result in all the input fields being empty, instead of showing their current values.

In the legacy forms on the site, the current form values are not included when the form is rendered, but in new designs (eg. As seen in the mocks at https://trello.com/c/KIU3NXPe/411-inline-editing-enhancement-for-forms) the forms show the current field values when they are initially displayed.

On this basis I would suggest that render should be passed the appstruct that was saved.